src/compiler/parser.ts

```diff
@@ -3480,7 +3518,7 @@ module ts {
                 case SyntaxKind.VoidKeyword:
                     return parseVoidExpression();
                 case SyntaxKind.LessThanToken:
-                    return parseTypeAssertion();
+                    return (sourceFile.jsxFactory && tryParse(() => parseJSXElement(true))) || parseTypeAssertion();

```

Comment:
I definitely see what you're saying.  However, i'm still very concerned here as the current approach means that *every* type assertion in a .tsx file will cause this potentially unbounded lookahead.  Once you're in an JSXChild context your scanner will keep on returning JSXText until it hits a ```{```, ```}```, ```<``` or ```</```.  And only ```}``` and ```</``` actually cause us to stop parsing text.  As such, it would be possible for us to scan huge swaths of code before finally deciding that it isn't a JSXElement, and trying again as a type assertion.  And, unfortunately, because you can easily have lots of code with nested type assertions, we'll be doing that lookahead over and over again.

Now: i totally get the concern for the case you've mentioned.  So there are approaches we can take to mitigate this issue and hopefully provide good perf, and better handling of code like what you've written above.  First, we would use the close-tag table as outlined above.  This would allow us to parse efficiently in the case where you had a correctly written close tag.   *However* we could then augment that slightly.  Specifically, if we run into a ```<tag>``` and we don't see a ```</tag>``` in our table, then we can look ahead a bit to see if what follows is a valid unary expression sequence.  If it isn't (say because you have ```<tag>foo bar```, then we go ahead and still parse it as an XML literal.

There is precedence because this is how we handle certain things like ```yield```.  Even if you're not in a generator function, but we see something like ```yield a```, then we realize "Hey! treating ```yield``` as an identifier here won't succeed, and will likely produce worse code.  So let's just parse out the actual yield-expression, and give a good message that it's not legal outside of a generator."  Such an approach could be viable here.  Now instead of unbounded lookahead in all cases, we do limited lookahead in the cases where we don't see a viable closing tag.